File: src/main/java/com/jozufozu/exnihiloomnia/common/registries/HeatSource.java
package com.jozufozu.exnihiloomnia.common.registries;

import com.google.gson.JsonObject;
import com.google.gson.JsonSyntaxException;
import com.jozufozu.exnihiloomnia.common.lib.LibRegistries;
import com.jozufozu.exnihiloomnia.common.registries.ingredients.WorldIngredient;
import net.minecraft.block.state.IBlockState;
import net.minecraft.util.JsonUtils;
import net.minecraftforge.registries.IForgeRegistryEntry;

public class HeatSource extends IForgeRegistryEntry.Impl<HeatSource>
{
    private int heatLevel;
    
    private WorldIngredient heatSource;
    
    public int getHeatLevel()
    {
        return heatLevel;
    }
    
    public boolean matches(IBlockState state)
    {
        return heatSource.test(state);
    }
    
    public static HeatSource deserialize(JsonObject object)
    {
        if (!object.has("source"))
        {
            throw new JsonSyntaxException("Heat source is missing block!");
        }
        
        HeatSource out = new HeatSource();
        
        out.heatLevel = JsonUtils.getInt(object, LibRegistries.HEAT);
        
        out.heatSource = WorldIngredient.deserialize(object.getAsJsonObject("source"));
        
        return out;
    }
}

File: src/main/kotlin/com/jozufozu/exnihiloomnia/common/registries/HeatSource.kt
package com.jozufozu.exnihiloomnia.common.registries

import com.google.gson.JsonObject
import com.google.gson.JsonSyntaxException
import com.jozufozu.exnihiloomnia.common.lib.LibRegistries
import com.jozufozu.exnihiloomnia.common.registries.ingredients.WorldIngredient
import com.jozufozu.exnihiloomnia.common.util.contains
import net.minecraft.block.BlockState
import net.minecraft.util.JSONUtils
import net.minecraftforge.common.crafting.CraftingHelper
import net.minecraftforge.registries.ForgeRegistryEntry

class HeatSource(
        val heatLevel: Int,
        val heatSource: WorldIngredient
) : ForgeRegistryEntry<HeatSource>() {

    fun matches(state: BlockState) = heatSource.test(state)

    companion object Serde {

        fun deserialize(heatSource: JsonObject): HeatSource? {
            if (LibRegistries.SOURCE !in heatSource) throw JsonSyntaxException("heat source is missing \"${LibRegistries.SOURCE}\"")

            if (!CraftingHelper.processConditions(heatSource, LibRegistries.CONDITIONS)) return null

            val heatLevel = JSONUtils.getInt(heatSource, LibRegistries.HEAT)

            RegistryLoader.pushCtx(LibRegistries.SOURCE)
            val source = WorldIngredient.deserialize(heatSource[LibRegistries.SOURCE])
            RegistryLoader.popCtx()

            return HeatSource(heatLevel, source)
        }
    }
}
